File: frontend/src/api/emulation.ts
import apiClient, { appendApiKey } from './client'
import { apiUrl } from './config'
import type {
  EmulationSession,
  EmulationStartRequest,
  EmulationExecResponse,
  EmulationPreset,
  EmulationPresetCreate,
  EmulationPresetUpdate,
  FirmwareService,
  SystemEmulationStartRequest,
  SystemCommandResponse,
} from '@/types'

export async function startEmulation(
  projectId: string,
  request: EmulationStartRequest,
  firmwareId?: string | null,
): Promise<EmulationSession> {
  const { data } = await apiClient.post<EmulationSession>(
    `/projects/${projectId}/emulation/start`,
    request,
    { params: firmwareId ? { firmware_id: firmwareId } : undefined },
  )
  return data
}

export async function deleteSession(
  projectId: string,
  sessionId: string,
): Promise<void> {
  await apiClient.delete(`/projects/${projectId}/emulation/${sessionId}`)
}

export async function stopEmulation(
  projectId: string,
  sessionId: string,
): Promise<EmulationSession> {
  const { data } = await apiClient.post<EmulationSession>(
    `/projects/${projectId}/emulation/${sessionId}/stop`,
  )
  return data
}

export async function execInEmulation(
  projectId: string,
  sessionId: string,
  command: string,
  timeout = 30,
): Promise<EmulationExecResponse> {
  const { data } = await apiClient.post<EmulationExecResponse>(
    `/projects/${projectId}/emulation/${sessionId}/exec`,
    { command, timeout },
  )
  return data
}

export async function listSessions(
  projectId: string,
): Promise<EmulationSession[]> {
  const { data } = await apiClient.get<EmulationSession[]>(
    `/projects/${projectId}/emulation/sessions`,
  )
  return data
}

export async function getSessionStatus(
  projectId: string,
  sessionId: string,
): Promise<EmulationSession> {
  const { data } = await apiClient.get<EmulationSession>(
    `/projects/${projectId}/emulation/${sessionId}/status`,
  )
  return data
}

export async function getSessionLogs(
  projectId: string,
  sessionId: string,
): Promise<string> {
  const { data } = await apiClient.get<{ logs: string }>(
    `/projects/${projectId}/emulation/${sessionId}/logs`,
  )
  return data.logs
}

export function buildEmulationTerminalURL(
  projectId: string,
  sessionId: string,
): string {
  const proto = window.location.protocol === 'https:' ? 'wss:' : 'ws:'
  const host = window.location.host
  return appendApiKey(
    `${proto}//${host}/api/v1/projects/${projectId}/emulation/${sessionId}/terminal`,
  )
}

// ── Emulation Presets ──

export async function listPresets(
  projectId: string,
): Promise<EmulationPreset[]> {
  const { data } = await apiClient.get<EmulationPreset[]>(
    `/projects/${projectId}/emulation/presets`,
  )
  return data
}

export async function createPreset(
  projectId: string,
  request: EmulationPresetCreate,
): Promise<EmulationPreset> {
  const { data } = await apiClient.post<EmulationPreset>(
    `/projects/${projectId}/emulation/presets`,
    request,
  )
  return data
}

export async function updatePreset(
  projectId: string,
  presetId: string,
  request: EmulationPresetUpdate,
): Promise<EmulationPreset> {
  const { data } = await apiClient.patch<EmulationPreset>(
    `/projects/${projectId}/emulation/presets/${presetId}`,
    request,
  )
  return data
}

export async function deletePreset(
  projectId: string,
  presetId: string,
): Promise<void> {
  await apiClient.delete(`/projects/${projectId}/emulation/presets/${presetId}`)
}

// ── System Emulation (FirmAE) ──

export async function startSystemEmulation(
  projectId: string,
  firmwareId: string,
  request?: SystemEmulationStartRequest,
): Promise<EmulationSession> {
  const { data } = await apiClient.post<EmulationSession>(
    `/projects/${projectId}/emulation/system`,
    request ?? {},
    { params: { firmware_id: firmwareId } },
  )
  return data
}

export async function getSystemEmulationStatus(
  projectId: string,
  sessionId: string,
): Promise<EmulationSession> {
  const { data } = await apiClient.get<EmulationSession>(
    `/projects/${projectId}/emulation/system/${sessionId}`,
  )
  return data
}

export async function getSystemEmulationServices(
  projectId: string,
  sessionId: string,
): Promise<FirmwareService[]> {
  const { data } = await apiClient.get<FirmwareService[]>(
    `/projects/${projectId}/emulation/system/${sessionId}/services`,
  )
  return data
}

export async function stopSystemEmulation(
  projectId: string,
  sessionId: string,
): Promise<void> {
  await apiClient.delete(`/projects/${projectId}/emulation/system/${sessionId}`)
}

export async function runCommandInFirmware(
  projectId: string,
  sessionId: string,
  command: string,
  timeout = 30,
): Promise<SystemCommandResponse> {
  const { data } = await apiClient.post<SystemCommandResponse>(
    `/projects/${projectId}/emulation/system/${sessionId}/command`,
    { command, timeout },
  )
  return data
}

export function buildSystemEmulationTerminalURL(
  projectId: string,
  sessionId: string,
  port: number,
): string {
  const proto = window.location.protocol === 'https:' ? 'wss:' : 'ws:'
  const host = window.location.host
  return `${proto}//${host}/api/v1/projects/${projectId}/emulation/system/${sessionId}/ws/${port}`
}

// ── Network Traffic Analysis ──

export interface NetworkCaptureResult {
  packet_count: number
  pcap_path: string
  size_bytes: number
  duration: number
}

export interface ProtocolBreakdown {
  protocol: string
  packet_count: number
  percentage: number
}

export interface NetworkConversation {
  src: string
  src_port: number
  dst: string
  dst_port: number
  protocol: string
  packet_count: number
  byte_count: number
}

export interface InsecureProtocolFinding {
  protocol: string
  port: number
  severity: string
  description: string
  evidence: string
  packet_count: number
}

export interface DnsQueryResult {
  domain: string
  query_type: string
  resolved_ips: string[]
}

export interface TlsInfoResult {
  server: string
  port: number
  version: string
  cipher_suites: string[]
}

export interface PcapAnalysis {
  total_packets: number
  protocol_breakdown: ProtocolBreakdown[]
  conversations: NetworkConversation[]
  insecure_findings: InsecureProtocolFinding[]
  dns_queries: DnsQueryResult[]
  tls_info: TlsInfoResult[]
}

export async function captureNetworkTraffic(
  projectId: string,
  sessionId: string,
  duration = 10,
  iface = 'eth0',
): Promise<NetworkCaptureResult> {
  const { data } = await apiClient.post<NetworkCaptureResult>(
    `/projects/${projectId}/emulation/system/${sessionId}/capture`,
    { duration, interface: iface },
  )
  return data
}

export async function analyzeNetworkTraffic(
  projectId: string,
  sessionId: string,
): Promise<PcapAnalysis> {
  const { data } = await apiClient.get<PcapAnalysis>(
    `/projects/${projectId}/emulation/system/${sessionId}/network-analysis`,
  )
  return data
}

export function getPcapDownloadUrl(
  projectId: string,
  sessionId: string,
): string {
  return apiUrl(`/api/v1/projects/${projectId}/emulation/system/${sessionId}/pcap`)
}

File: backend/app/schemas/emulation.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class PortForward(BaseModel):
    host: int
    guest: int


class EmulationStartRequest(BaseModel):
    mode: Literal["user", "system"]
    binary_path: str | None = None
    arguments: str | None = None
    port_forwards: list[PortForward] = []
    kernel_name: str | None = None
    init_path: str | None = None  # Override /sbin/init (e.g., "/bin/sh" or "/bin/busybox")
    pre_init_script: str | None = None  # Shell script to run before firmware init (e.g., start cfmd, set up LD_PRELOAD)
    stub_profile: Literal["none", "generic", "tenda"] | None = None  # Stub library profile for system-mode emulation


class EmulationSessionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    firmware_id: uuid.UUID
    mode: str
    status: str
    architecture: str | None
    binary_path: str | None
    arguments: str | None
    port_forwards: list[dict] | None
    error_message: str | None
    logs: str | None
    started_at: datetime | None
    stopped_at: datetime | None
    created_at: datetime


class EmulationExecRequest(BaseModel):
    command: str
    timeout: int = Field(default=30, ge=1, le=120)
    environment: dict[str, str] | None = None


class EmulationExecResponse(BaseModel):
    stdout: str
    stderr: str
    exit_code: int
    timed_out: bool


# ── Emulation Presets ──


class PortForwardItem(BaseModel):
    host: int
    guest: int


class EmulationPresetCreate(BaseModel):
    name: str = Field(..., max_length=255)
    description: str | None = None
    mode: Literal["user", "system"]
    binary_path: str | None = None
    arguments: str | None = None
    architecture: str | None = None
    port_forwards: list[PortForwardItem] = []
    kernel_name: str | None = None
    init_path: str | None = None
    pre_init_script: str | None = None
    stub_profile: Literal["none", "generic", "tenda"] = "none"


class EmulationPresetUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    description: str | None = None
    mode: Literal["user", "system"] | None = None
    binary_path: str | None = None
    arguments: str | None = None
    architecture: str | None = None
    port_forwards: list[PortForwardItem] | None = None
    kernel_name: str | None = None
    init_path: str | None = None
    pre_init_script: str | None = None
    stub_profile: Literal["none", "generic", "tenda"] | None = None


class EmulationPresetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    name: str
    description: str | None
    mode: str
    binary_path: str | None
    arguments: str | None
    architecture: str | None
    port_forwards: list[dict] | None
    kernel_name: str | None
    init_path: str | None
    pre_init_script: str | None
    stub_profile: str
    created_at: datetime
    updated_at: datetime


# ── System Emulation (FirmAE) ──


class SystemEmulationStartRequest(BaseModel):
    brand: str = "unknown"
    timeout: int = Field(default=600, ge=60, le=3600)


class FirmwareServiceResponse(BaseModel):
    port: int
    protocol: str
    service: str
    host_port: int | None = None
    url: str | None = None


class SystemEmulationStatusResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    firmware_id: uuid.UUID
    mode: str
    status: str
    architecture: str | None
    binary_path: str | None
    arguments: str | None
    port_forwards: list[dict] | None
    error_message: str | None
    logs: str | None
    started_at: datetime | None
    stopped_at: datetime | None
    created_at: datetime
    # System emulation fields
    discovered_services: list[dict] | None = None
    system_emulation_stage: str | None = None
    kernel_used: str | None = None
    firmware_ip: str | None = None
    nvram_state: dict | None = None
    idle_since: datetime | None = None
    pcap_path: str | None = None


class SystemCommandRequest(BaseModel):
    command: str
    timeout: int = Field(default=30, ge=1, le=120)


class SystemCommandResponse(BaseModel):
    stdout: str
    stderr: str
    exit_code: int


class NetworkCaptureRequest(BaseModel):
    duration: int = Field(default=10, ge=1, le=120)
    interface: str = "eth0"


class NetworkCaptureResponse(BaseModel):
    packet_count: int
    pcap_path: str
    size_bytes: int
    duration: int


class NvramResponse(BaseModel):
    nvram: dict[str, str]


# ── Pcap Analysis ──


class ProtocolBreakdownResponse(BaseModel):
    protocol: str
    packet_count: int
    percentage: float


class ConversationResponse(BaseModel):
    src: str
    src_port: int
    dst: str
    dst_port: int
    protocol: str
    packet_count: int
    byte_count: int


class InsecureProtocolResponse(BaseModel):
    protocol: str
    port: int
    severity: str
    description: str
    evidence: str
    packet_count: int


class DnsQueryResponse(BaseModel):
    domain: str
    query_type: str
    resolved_ips: list[str]


class TlsInfoResponse(BaseModel):
    server: str
    port: int
    version: str
    cipher_suites: list[str]


class PcapAnalysisResponse(BaseModel):
    total_packets: int
    protocol_breakdown: list[ProtocolBreakdownResponse]
    conversations: list[ConversationResponse]
    insecure_findings: list[InsecureProtocolResponse]
    dns_queries: list[DnsQueryResponse]
    tls_info: list[TlsInfoResponse]
